### ISSUE TITLE
Cross-validation or validation split for ImageDataGenerator/flow_from_directory?

### ISSUE THREAD

**Issue by hanfeisun** [fchollet/keras#6452]
I'm working on a classifier. Currently my data structure looks like this:

```
data/
        dogs/
            dog001.jpg
            dog002.jpg
            ...
        cats/
            cat001.jpg
            cat002.jpg
```

I want to use flow_from_directory to do the training and validation.. But it is a bit clumsy to move the files into a different directory each time. Is there any option that can help that?

**Comment by DanMossa** [comment]
Still looking for something like

**Comment by oxydron** [comment]
Extremely relevant issue.

**Comment by atoaster** [comment]
Super late response, but for whoever finds this: `flow_from_directory()` is a generator which means it doesn't have knowledge of the entire directory, so can't create a training and validation set.

I was reading another post where they made a script that splits the folders into training/testing splits here:
```
def arrange_dataset(directory, file_extension='*.jpg'):
		# Directory should be the dataset directory.
		if not os.path.exists(directory):
			return 0
		cur_dir = os.getcwd()
		os.chdir(directory)
		work_dir = os.getcwd()
		file_list = glob.glob('*.jpg')
		train_dir = os.path.abspath('train')
		test_dir = os.path.abspath('test')

		if not os.path.exists(train_dir):
			os.mkdir(train_dir)
		if not os.path.exists(test_dir):
			os.mkdir(test_dir)

		os.chdir(cur_dir)

		# set init label
		init_label = file_list[0][0:4]
		current_list = []
		for file in file_list:
			label = file[0:4]
			if label == init_label:
				if not os.path.exists(os.path.join(train_dir, label)):
					os.mkdir(os.path.join(train_dir, label))
				if not os.path.exists(os.path.join(test_dir, label)):
					os.mkdir(os.path.join(test_dir, label))
				current_list.append(file)
			if (label != init_label) or (file == file_list[-1]):
				partial_train, partial_test = ImageReadUtils.arrange_dataset_2(current_list)
				for train_file in partial_train:
					os.rename(os.path.join(work_dir, train_file), os.path.join(train_dir, init_label, train_file))
				for test_file in partial_test:
					os.rename(os.path.join(work_dir, test_file), os.path.join(test_dir, init_label, test_file))
				current_list[:] = []
				init_label = label
				current_list.append(file)

def arrange_dataset_2(file_list, train_split=0.8):
		if len(file_list) == 2:
			train_split = 0.5
		elif len(file_list) == 3:
			train_split = 0.66
		elif len(file_list) == 4:
			train_split = 0.75
		random_set = np.random.permutation(len(file_list))
		train_list = random_set[:round(len(random_set)*0.8)]
		test_list = random_set[-(len(file_list) - len(train_list))::]
		train_images = []
		test_images = []
		for index in train_list:
			train_images.append(file_list[index])
		for index in test_list:
			test_images.append(file_list[index])
		return train_images, test_images
```
source:

https://github.com/keras-team/keras/issues/5862

**Comment by Metete** [comment]
@atoaster As far as I observed this code snippet is for hold-out split not for cross validation. So far I could not find any example with fit_generator and cross-validation using keras.

**Comment by cs-karthik** [comment]
Yes there is a inbuilt way to do it using Keras.
train_datagen = ImageDataGenerator(
    rescale=1. / 255,
    shear_range=0.2,
    zoom_range=0.2,
    validation_split=0.2)#for validation

train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='categorical',
    subset='training',seed=42) # set as training data

validation_generator = train_datagen.flow_from_directory(
    train_data_dir,#same train directory
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='categorical',
    subset='validation',seed=42) # set as validation data

**Comment by onietosi** [comment]
Note that this provides a _single_ validation split and does not perform cross-validation.

**Comment by m1raz3nt** [comment]
To anyone, who bumped into this problem: to the date, at which this answer was posted - there's no (at least, relatively) simple out-of-the-box solution in my opinion and deciding by the result of my own searches.

The only solution, that I came up with, resolving similar problem in my project, was to make partitions in my dataset, with number of partitions equal to number of folds, and saving them as dictionary with number of partition as a key and file paths list as a value for partition. After that, you still have to sort your files into class folders for train and validation subsets respectively. Thus, you still require two generators, validation generator processes one of the partitions, train - processes the rest of them.

**Comment by Sicily-F** [comment]
Hi everyone!!! I'm struggling with a similar issue. The best thing I could come up with was using the splitfolders package to randomly split your data 10 times shall we say, i.e. into 10 folds, and then run your generators on each new 'fold' dataset - although, I'm not sure it makes sense for me to split my original dataset say 10 times because I have a relatively small dataset, on average 150 photos per class, for 36 classes.  I saw what looked like a good solution here, but couldn't get it to work: https://datascience.stackexchange.com/questions/72372/k-fold-crossvalidation-in-tensorflow-when-using-flow-from-directory-for-image-re

Anyone have any more recent success?
